File: plugins/plugin-slack/rust/src/providers/member_list.rs
//! Member list provider for Slack.

use crate::service::SlackService;
use serde::{Deserialize, Serialize};

/// Provider name
pub const MEMBER_LIST_PROVIDER: &str = "slackMemberList";

/// Provider description
pub const MEMBER_LIST_DESCRIPTION: &str = "Provides information about members in the current Slack channel";

/// Member info
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemberInfo {
    pub id: String,
    pub name: String,
    pub display_name: String,
    pub is_bot: bool,
    pub is_admin: bool,
}

/// Member list data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemberListData {
    pub channel_id: String,
    pub channel_name: String,
    pub member_count: usize,
    pub members: Vec<MemberInfo>,
    pub has_more_members: bool,
}

/// Member list result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemberListResult {
    pub data: MemberListData,
    pub text: String,
}

/// Get member list (simplified - would need full implementation with conversations.members API)
pub async fn get_member_list(
    service: &SlackService,
    channel_id: &str,
) -> Result<MemberListResult, String> {
    let channel = service
        .get_channel(channel_id)
        .await
        .map_err(|e| e.to_string())?;

    let channel_name = channel.name.clone();
    let member_count = channel.num_members.unwrap_or(0) as usize;

    // Note: Full implementation would call conversations.members and users.info
    // This is a simplified version
    let response_text = format!(
        "Members in #{} ({} total members)",
        channel_name, member_count
    );

    Ok(MemberListResult {
        data: MemberListData {
            channel_id: channel_id.to_string(),
            channel_name,
            member_count,
            members: vec![], // Would be populated by conversations.members
            has_more_members: false,
        },
        text: response_text,
    })
}

File: plugins/plugin-slack/src/providers/memberList.ts
import type { IAgentRuntime, Memory, Provider, State } from "@elizaos/core";
import { validateActionKeywords, validateActionRegex } from "@elizaos/core";
import type { SlackService } from "../service";
import { getSlackUserDisplayName, ServiceType } from "../types";

/**
 * Provider for retrieving Slack channel member information.
 */
export const memberListProvider: Provider = {
  name: "slackMemberList",
  description:
    "Provides information about members in the current Slack channel",
  dynamic: true,
  contexts: ["messaging", "connectors"],
  contextGate: { anyOf: ["messaging", "connectors"] },
  cacheScope: "conversation",
  roleGate: { minRole: "ADMIN" },
  relevanceKeywords: [
    "slackmemberlist",
    "memberlistprovider",
    "plugin",
    "slack",
    "status",
    "state",
    "context",
    "info",
    "details",
    "chat",
    "conversation",
    "agent",
    "room",
    "channel",
  ],
  get: async (runtime: IAgentRuntime, message: Memory, state: State) => {
    const __providerKeywords = [
      "slackmemberlist",
      "memberlistprovider",
      "plugin",
      "slack",
      "status",
      "state",
      "context",
      "info",
      "details",
      "chat",
      "conversation",
      "agent",
      "room",
      "channel",
    ];
    const __providerRegex = new RegExp(
      `\\b(${__providerKeywords.join("|")})\\b`,
      "i",
    );
    const __recentMessages =
      (state?.recentMessagesData as Memory[] | undefined) ?? [];
    const __isRelevant =
      validateActionKeywords(message, __recentMessages, __providerKeywords) ||
      validateActionRegex(message, __recentMessages, __providerRegex);
    if (!__isRelevant) {
      return { text: "" };
    }

    // If message source is not slack, return empty
    if (message.content.source !== "slack") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const room = state.data?.room ?? (await runtime.getRoom(message.roomId));
    if (!room?.channelId) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const slackService = runtime.getService(ServiceType.SLACK) as SlackService;
    if (!slackService?.client) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const channelId = room.channelId;

    // Get channel members
    try {
      const membersResult = await slackService.client.conversations.members({
        channel: channelId,
        limit: 100,
      });

      const memberIds = membersResult.members || [];

      if (memberIds.length === 0) {
        return {
          data: {
            channelId,
            memberCount: 0,
            members: [],
          },
          values: {
            memberCount: 0,
          },
          text: "No members found in this channel.",
        };
      }

      // Get user info for each member (limited to first 20 for performance)
      const memberLimit = 20;
      const limitedMemberIds = memberIds.slice(0, memberLimit);
      const members: Array<{
        id: string;
        name: string;
        displayName: string;
        isBot: boolean;
        isAdmin: boolean;
      }> = [];

      for (const memberId of limitedMemberIds) {
        const user = await slackService.getUser(memberId);
        if (user) {
          members.push({
            id: user.id,
            name: user.name,
            displayName: getSlackUserDisplayName(user),
            isBot: user.isBot,
            isAdmin: user.isAdmin || user.isOwner,
          });
        }
      }

      // Get channel info for name
      const channel = await slackService.getChannel(channelId);
      const channelName = channel?.name || channelId;

      // Format member list
      const botUserId = slackService.getBotUserId();
      const memberDescriptions = members.map((m) => {
        const tags: string[] = [];
        if (m.id === botUserId) tags.push("this bot");
        if (m.isBot && m.id !== botUserId) tags.push("bot");
        if (m.isAdmin) tags.push("admin");
        const tagStr = tags.length > 0 ? ` (${tags.join(", ")})` : "";
        return `- ${m.displayName} (@${m.name})${tagStr}`;
      });

      const truncationNote =
        memberIds.length > memberLimit
          ? `\n\n(Showing ${memberLimit} of ${memberIds.length} total members)`
          : "";

      const responseText = `Members in #${channelName}:\n${memberDescriptions.join("\n")}${truncationNote}`;

      return {
        data: {
          channelId,
          channelName,
          memberCount: memberIds.length,
          members,
          hasMoreMembers: memberIds.length > memberLimit,
        },
        values: {
          channelId,
          channelName,
          memberCount: memberIds.length,
        },
        text: responseText,
      };
    } catch (error) {
      return {
        data: {
          channelId,
          memberCount: 0,
          members: [],
          error: error instanceof Error ? error.message : String(error),
        },
        values: {
          channelId,
          memberCount: 0,
          slackMembersAvailable: false,
        },
        text: "Slack member list unavailable.",
      };
    }
  },
};

export default memberListProvider;
